software/contrib/egressus_melodiam.py

```diff
@@ -0,0 +1,971 @@
+There are both linear and non-linear interpolation functions that create various smooth and no-so-smooth shapes.
+
+Each slew shape function generates an array of values (samples) between two given points.
+Samples are output on the CV outputs based on the list of pre-computed interpolated values (sample buffers)
+A value from the sample buffer is retrieved using a Python generator function (self.slewGenerator()).
+One instance of the generator function for each each slew shape exists in a list: self.slewGeneratorObjects[]
+
+A new sample buffer (array of interpolated values) is created at each clock step.
+The number of samples required in each sample buffer (one for each output) is calculated at each clock step or if
+the clock rate or output division changes.
+
+The timing and associated hanlding of input clock interrupts (clocks to din) is not perfect - python/micropython is not
+known to be great at sub-second timing!
+Therefore, at the end of each clock step the algorithm tunes itself to work around sample buffer overruns or underruns.
+If there is a buffer underrun (not enough samples in the buffer), the previous output voltage (sample) is used until the algorithm
+catches back up with itself.
+
+slewGeneratorObjects[] contains 6 object (one for each output)
+Each slewGeneratorObjects[] object is a reference to a copy of the self.slewGenerator() function.
+The self.slewGenerator() function receives a buffer filled with samples and yields one sample each time it is called using next()
+
+In order to maintain the best balance of smooth waves and Rpi pico memory usage, an algorithm is used to vary the 
+sample rate automatically based on the selected output division.
+This causes the sample rate to be at minium when there is a slow clock and high output division.
+Conversely, when there is a fast clock and low output division a higher sample rate is used to avoid unwanted steps
+often caused by under-samping a wave form.
+
+"""
+# Minimum allowed time between incoming 16th note clocks. Smaller values are too fast for poor ole micropython to keep up
+MIN_CLOCK_TIME_MS = 50
+
+# Maximum allowed time between incoming 16th note clocks. This is 2 BPM, larger value causes memory issues unless the next two vals are halfed
+MAX_CLOCK_TIME_MS = 3750
+
+# Max sample rate and clock divisor to keep required buffer sizes within memory limits
+MAX_SAMPLE_RATE = 32
+MAX_OUTPUT_DENOMINATOR = 8
+
+# Calculate maximum sample buffer size required
+SLEW_BUFFER_SIZE_IN_SAMPLES = int(
+    (MAX_CLOCK_TIME_MS / 1000) * MAX_SAMPLE_RATE * MAX_OUTPUT_DENOMINATOR
+)
+
+# Keep these in for future debugging during feature expansion
+DEBUG_VERBOSE = 3
+DEBUG_LOTS = 2
+NO_DEBUG = 0
+DEBUG_MODE = NO_DEBUG
+
+# Attempt to avoid knob hysteresis
+KNOB_CHANGE_TOLERANCE = 0.999
+
+# put in a try catch block so it does not need the latest europi firmware version
+try:
+    MAX_CV_VOLTAGE = min(5, europi_config["max_output_voltage"])

```

Comment:
Thanks for making this optional.

I found that having a voltage range higher than 5V required a lot of attenuation and because of the wide range (0-10V for example) it was much harder to dial in an attenuation value using a pot than it was with a 5V range - i.e. a fractional pot movement caused too much of a change which might be frustrating for some (it was for me).

In terms of compatibility with other modules - I agree that restricting it to less than a globally configured maximum may cause an issue with some modules, but most that I checked use 0-5V or -5 to +5 (which without using an offset in the CV path would be difficult for users to control with a 10V CV range).

I think the compromise to provide a reasonable maximum value in the firmware and to allow script writers to deviate from this in their own scripts as they see fit strikes the right balance - this works well for script writers that want to use less than the value in the firmware.

What might be a concern though is if it difficult for script writers to deviate from the defaults using a higher voltage than that configured in the firmware - for legitimate reasons. For example, some modules require a 10V gate (pulsar-23). If I understand the most recent europi config stuff correctly, if a script is using cv.on() and requires 10V gates, then the script writer would need to create their own config json file as the value in the firmware has a default of 5V - this seems cumbersome. However, if this is a rare edge case (it might well be) then creating a new json file as a workaround is sorta OK. If I am mis-understanding the documentation though, please let me know - I would be very happy to learn that sending 10V gates using cv.on() is easier than I think it is!

Reply:
I think that based on this and the comment below, we haven't made the purpose of the config settings as clear as we probably should. 

Essentially the config JSON file is a user editable file **not** a script writer editable file. A `config_EuroPiConfig.json` file should never be included as a file change in a PR.

I think I'm failing to see your reasoning as valid from the point of a *script writer* even though it's completely valid from the point of a *user*. The modules you are likely to use with this script accept 0-5V, but for me I know that I'm going to plug it into my Befaco offset module and then use it to control Beads' ±5V inputs. I cannot do this if the script imposes a strict limit separate to the configured value, and I think it is confusing that if a user were to make a specific setting for maximum CV voltage based on their needs, this script would not respect it.

If you want to send 10V gates using `cv.on()`, the only thing you need to do is create the file `config/config_EuroPiConfig.json` on your Pico after loading the script and firmware, with the contents:
```
{
    "gate_voltage": 10
}
```
In future we are planning to make a contrib script which allows you to easily interactively edit the config values without needing to use Thonny/any IDE. 

I am aware that it seems a little cumbersome, but I think that the end result of *every* user being able to configure *every* script at once to use the range they need is a great thing, and feels like it fits in with the accessible and DIY nature of the project as a whole

Reply:
Thanks for taking the time to explain the newly released config feature. I think I get it. I also appreciate your differentiation between script writers and script users.

It also good to learn that script users will be able to use a UI at some point to configure these values.

Would you mind reviewing my proposed plan/understanding to see if this is aligned with the current config solution please?

1. All scripts should use global config values in scripts
2. If a script user would like to deviate from the global config values they should create a config json (in Thonny now, in the UI in the future)
3. As a script writer I should not submit a PR with a config file, instead I should add some lines to the MD file explaining to script users how to adjust config values should they want to.

Reply:
Exactly. The idea is that users don't need to know anything at all about each script's inner workings, but can trust that if they add values to the config file, the scripts will respect that and thus their module will work in their system, with the voltages they want it to output.